rust/lance/src/io/exec/scalar_index.rs

```diff
@@ -257,29 +257,76 @@ impl MaterializeIndexExec {
+            // bitmap.  This node only runs on indexed fragments and any fragments that were deleted
+            // when the index was trained will still be deleted when the index is queried.
+            PreFilter::create_deletion_mask(dataset.clone(), fragment_bitmap)
+        });
         let mask = if let Some(prefilter) = prefilter {
             let (mask, prefilter) = futures::try_join!(mask, prefilter)?;
             mask.also_block((*prefilter).clone())
         } else {
             mask.await?
         };
-        let ids = match (mask.allow_list, mask.block_list) {
+        let span = debug_span!("make_ids");
+        let ids = span.in_scope(|| match (mask.allow_list, mask.block_list) {
             (None, None) => FragIdIter::new(fragments).collect::<Vec<_>>(),
-            (Some(allow_list), None) => FragIdIter::new(fragments)
-                .filter(|row_id| allow_list.contains(*row_id))
-                .collect(),
+            (Some(allow_list), None) => {
+                let allowed_ids = allow_list.row_ids();
+                if let Some(allowed_ids) = allowed_ids {
+                    let frag_id_filter = fragments
+                        .iter()
+                        .map(|frag| frag.id as u32)
+                        .collect::<HashSet<_>>();
+                    allowed_ids
+                        .filter_map(|addr| {
+                            if frag_id_filter.contains(&addr.fragment_id()) {
+                                Some(u64::from(addr))
+                            } else {
+                                None
+                            }
+                        })
+                        .collect::<Vec<_>>()
+                } else {
+                    FragIdIter::new(fragments)
+                        .filter(|row_id| allow_list.contains(*row_id))
+                        .collect()
+                }
+            }
             (None, Some(block_list)) => FragIdIter::new(fragments)
                 .filter(|row_id| !block_list.contains(*row_id))
                 .collect(),
-            (Some(allow_list), Some(block_list)) => FragIdIter::new(fragments)
-                .filter(|row_id| !block_list.contains(*row_id) && allow_list.contains(*row_id))
-                .collect(),
-        };
+            (Some(allow_list), Some(block_list)) => {
+                let allowed_ids = allow_list.row_ids();
+                if let Some(allowed_ids) = allowed_ids {
+                    let frag_id_filter = fragments
+                        .iter()
+                        .map(|frag| frag.id as u32)
+                        .collect::<HashSet<_>>();
+                    allowed_ids
+                        .filter_map(|addr| {
+                            let frag_id = addr.fragment_id();
+                            let row_id = u64::from(addr);
+                            if !block_list.contains(row_id) && frag_id_filter.contains(&frag_id) {
```

Comment:
Similarly here, I think it might be more efficient to operate on the TreeMaps themselves. However, to actually make this one work, I think you'd have to implement a set difference operation on them, so might just leave a TODO for that.

Reply:
I can still filter the allow list I think?  Let me know if the updated seems good.  Not sure what you mean by needing a set difference?  Unless you're thinking of combining the allow_list and the block_list?